docsrc/interactivity.nim

```diff
@@ -253,4 +253,53 @@ And here the problem lies: `counter_2` is generated both times we compile the bl
 The solution is stated above: don't name multiple separate variables the same in a `nbKaraxCode` or `nbJsFromCodeOwnFile` block! 
 This isn't a problem for the other nbJs blocks luckily. 
 """
-nbSave
\ No newline at end of file
+nbText: hlMd"""
+### nbHappyxCode
+HappyX is an emerging alternative to Jester (on the back-end) and Karax(on the front end). It aims to streamline the syntax for writing full-stack applications and to allow more flexibility in its single page applications, which use a routing mechanism to switch between different pages for the app. It is being actively developed and some of the syntax for the DSL may change, so the introduction will be brief.
+
+The system for HappyX in nimib is analogous to the system for Karax. Note the parts of a typical Karax code block.
+
+```nim
+  template karaxExample =
+    let x = 3.14
+    nbKaraxCode(x):
+      var message = "Pi is roughly " & $x
+      karaxHtml:
+        p:
+          text message
+        button:
+          text "Click me!"
+          proc onClick() =
+            message = "Poof! Gone!"
+```
+Here's how it changes for HappyX:
+
+"""
+nimibCode:
+  template happyxExample =
+    let x = 3.14
+    nbHappyxCode(x):
+      var message = remember fmt"pi is roughly {x}"
+      happyxRoutes:
+        "/":
```

Comment:
is this `"/"` necessary or we can have a template which is just `happyxroot`? I do not really understand how happyx mixes the server and client side stuff (have not really studied how it works), I understand that routes are fundamental server side, but on the client side and in nimib should we really support them? shouldn't this be always the root in nimib? maybe there are some other use cases I do not know about (like supporting url parameters, which would be cool).

Reply:
It would be good practice even if having root as `"/"` were not necessary. `appRoutes` declares routes --  By default we want to always start by loading the content at `#/`. Since we could  alsohave other routes, so starting a `"/"` block keeps you mentally prepared to write the others with the same indentation. We could adapt the API to allow for a simplified option for the syntax but since nimib plays nice with the **front-end routes** of happyx, it's great to have the ability to use them in the `nbHappyxCode `. This serves most of your needs for switching between very different versions of the DOM:

```
     let x = 3.14
    nbHappyxCode(x):
      var message = remember fmt"pi is roughly {x}"
      happyxRoutes:
        "/":
          p:
           {message}
          tButton:
            "Click me!"
            @click(
              message.set("gonezo"))
          tButton:
            "Change world"
            @click(
              route("orange"))
        "orange":
          tDiv(style="color: orange;"):
            "Welcome to orange world."
            tButton:
              "Go back"
              @click(
                route("/"))
```
As you might expect, there are ways to avoid using routes on the front end if you need to, but they can be tedious and error-prone in practice. Imagine creating custom components and templates to manage a complete change of the DOM along with updates to relevant variables in Karax. Routes also give nice feedback to the user in the URL and allow them to bookmark the page with some of the intended state -- note that you can use the `...#orange` URL to navigate directly to orange world. I'm not sure of the roadmap toward supporting url parameters on the back-end.

Since happyx already supports the routing syntax on the front-end and back-end, why not use it? This will make working with the full stack more streamlined, and is great for something like switching between a thumbnail index and an in-depth view of some isolated part.  Nimib at the very least should be able to demonstrate what we can do with happyX. But nbKaraxCode and nbHappyxCode are useful precisely because sometimes we need to get the formatting and behavior 'verbatim' from the way it is implemented in another SPA.

Reply:
I do see some potential collisions in the routing of multiple nbHappyxCode instances. I haven't reached that use case yet, and for me it is not critical to have it. A practical rule of using 1 SPA app per webpage in development helps to keep the code readable and avoid errors, in my opinion.

Reply:
ah ok, I did not realize that frontend routes in happyx are implemented using html fragment identifiers. TIL this is called hash-based routing. having understood this, I think it is definitely appropriate to leave `nbHappyxCode` as it is.